### ISSUE TITLE
Style/RedundantSelf cop doesn't understand classes that use method_missing

### ISSUE THREAD

**Issue by ifreecarve** [bbatsov/rubocop#1933]
The RedundantSelf cop doesn't check to see whether the uses of `self.` refer to something that does not exist in the class.  For example, classes that extend `ostruct` would need the `self.` in order to trigger the proper behavior -- the method_missing call which creates the necessary data structures.

Example:

```ruby

require 'ostruct'

class MyStruct < OpenStruct

  def initialize
    self.right_way = OpenStruct.new
    @wrong_way = OpenStruct.new
  end  

end

ms = MyStruct.new
ms.right_way.test1 = "this functions correctly, but rubocop complains about it"
ms.wrong_way.test2 = "in `initialize': undefined method `test2=' for nil:NilClass (NoMethodError)"
```

The proper solution would seem to be ignoring instances of `self.` that do not refer to defined class methods or variables.

**Comment by alexdowad** [comment]
`RedundantSelf` doesn't warn about assignment methods like `#right_way=`, so this is a non-issue. I recommend it should be closed.

**Comment by ifreecarve** [comment]
Is there a way you'd recommend writing the above code so that it can both work properly and pass the rubocop check?

**Comment by alexdowad** [comment]
```
16:55 ~prog/Ruby/rubocop % ruby -Ilib bin/rubocop test2.rb                 [more_fixes]
warning: parser/current is loading parser/ruby22, which recognizes
warning: 2.2.3-compliant syntax, but you are running 2.2.2.
warning: please see https://github.com/whitequark/parser#compatibility-with-ruby-mri.
Inspecting 1 file
C

Offenses:

test2.rb:1:1: C: Missing utf-8 encoding comment.
require 'ostruct'
^
test2.rb:3:1: C: Missing top-level class documentation comment.
class MyStruct < OpenStruct
^^^^^
test2.rb:4:1: C: Extra empty line detected at class body beginning.
test2.rb:8:6: C: Trailing whitespace detected.
  end  
     ^^
test2.rb:9:1: C: Extra empty line detected at class body end.
test2.rb:13:22: C: Prefer single-quoted strings when you don't need string interpolation or special symbols.
ms.right_way.test1 = "this functions correctly, but rubocop complains about it"
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
test2.rb:14:81: C: Line is too long. [98/80]
ms.wrong_way.test2 = "in `initialize': undefined method `test2=' for nil:NilClass (NoMethodError)"
                                                                                ^^^^^^^^^^^^^^^^^^
```

**Comment by alexdowad** [comment]
I would recommend you:

- Add a UTF-8 encoding comment.
- Add a comment for what `MyStruct` is about.
- Remove the excess whitespace and extra empty lines.
- Use a single-quoted string as suggested.
- Split the long string over two lines rather than one.

Then you are good to go!

**Comment by ifreecarve** [comment]
Hmm, it appears that the example code no longer triggers the RedundantSelf cop, and I'm not sure what changed.
